[DDP, dynamo] setup/unwind `autocast` and `set_grad_enabled` across `fx.passes.split_module` module splits

jon-chuang opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/111794 (fixes the repro)

DDPOptimizer splits modules. We need to setup/unwind global states (autocast, grad_enabled) for each split, as this affects downstream compilation.

TODO:
- [x] add small repro test
    - test must ensure that the correct autocast/grad is in effect for every split
    - can use a smaller `bucket_bytes_cap` to induce more graph splits for smaller graphs
- [x] stack this PR's style changes
- [x] improve logging 
- [ ] possibly convert asserts into appropriate errors
cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

jon-chuang committed:
done

complete

complete

done

jon-chuang commented on torch/fx/passes/split_module.py:
```diff
@@ -195,52 +277,68 @@ def record_cross_partition_use(
+                new_node = partition.graph.create_node(
+                    op=node.op,
+                    target=node.target,
+                    args=tuple(arg for arg in node.args),
```
can we rely on canonicalization of kwargs to args? Seems like yes, but why?

jon-chuang commented:
@wconstab please advise if it is worth landing this since I understand the [long-term plan](https://github.com/pytorch/pytorch/issues/109774) is to get rid of `DDPOptimizer` by moving the bucketing logic and allreduce into Inductor.

I don't think that approach would suffer from the same issues as `split_module` wrt to the context managers.

jon-chuang commented on torch/fx/passes/split_module.py:
```diff
@@ -162,27 +165,106 @@ def construct_graph(
+
+    # Super Nodes are nodes which by their global state effects,
+    # "taint" all downstream nodes while they are active.
+    SUPER_NODE_TARGETS = [
```
Maybe call it `GLOBAL_STATE_NODES`?

wconstab commented on torch/fx/passes/split_module.py:
```diff
@@ -28,17 +31,17 @@ def __repr__(self) -> str:
 
 
-# Creates subgraphs out of main graph
+# Creates subgraphs out of main graphf
```
typo

wconstab commented on torch/fx/passes/split_module.py:
```diff
@@ -162,27 +165,106 @@ def construct_graph(
+    # 2. subsequent regions: we insert the set_grad at the beginning
+    grad_regions: OrderedDict[Node, Set[int]] = OrderedDict()
+
+    # For autocast regions:
```
why is autocast different than grad? (why turn autocast off at the end of first region)

jon-chuang replied:
Autocast behaves differently - it is context manager style i.e. exiting at the call to _exit_autocast - so we want to handle nested.

Grad is just mutating the global state. So it owns the entirety of the subsequent region until next set grad.

jon-chuang replied:
Technically only 1 autocast needs to own an intermediate region, but to simplify the reasoning we allow multiple ownership.

wconstab reviewed (commented):
i did a brief/partial review and i think generally this looks worth doing; its really not DDPOptimizer specific, its more 'fixing fx split module' to work with more of torch features.  So from that perspective i'd say its worth landing but needs careful review and thorough testing.  Would be good to get input from @ezyang @bdhirsh @Chillee 

Also as a general note, you could split out the changes that are more 'renaming for clarity' to a separate PR, that could be landed already imo (and make the remainder of this PR smaller to read)

jon-chuang commented on torch/fx/passes/split_module.py:
```diff
@@ -195,52 +277,68 @@ def record_cross_partition_use(
-        torch.fx.graph.map_arg(
-            node.kwargs, lambda def_node: record_cross_partition_use(def_node, node)
-        )  # noqa: B950
+        if node.target not in SUPER_NODE_TARGETS:
```
Add comment: do not capture cross-partition dependencies for global state nodes as their setup/unwind will be inlined into the graph.

jon-chuang commented on torch/fx/passes/split_module.py:
```diff
@@ -162,27 +165,106 @@ def construct_graph(
+                autocast_regions[node] = set({split_callback(node)})
+                autocast_exits[node] = None
+            elif node.target == torch.amp._exit_autocast:
+                assert len(node.args) == 1
```
For correctness sake, you still need to write 

`autocast_regions[node.args[0]].add(split_callback(node))`

ezyang commented:
Can you post some examples of before graph, and what the graph looks like after splitting? (These would make good tests, but in lieu of that, some debug prints would help review too.)

jon-chuang commented:
Right, there's an example graph here (the final submodule) https://github.com/pytorch/pytorch/issues/111794#issuecomment-1782273063

Will post a before and after soon (+ easy repro / tests).

ezyang commented:
bumping this. It's important to see the before, because otherwise just seeing the submodule in a vacuum won't tell me if it's right or not

jon-chuang commented:
@ezyang I added a test.

split modules fx graphs before:
```
[rank0]:[2023-11-09 12:39:12,482] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] 
[rank0]:[2023-11-09 12:39:12,482] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---submod_0 graph---
[rank0]:[2023-11-09 12:39:12,482] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] graph():
[rank0]:[2023-11-09 12:39:12,482] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l_inputs_ : torch.Tensor [num_users=1] = placeholder[target=l_inputs_]
[rank0]:[2023-11-09 12:39:12,482] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %_set_grad_enabled : [num_users=0] = call_function[target=torch._C._set_grad_enabled](args = (False,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,482] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_0 : [num_users=1] = call_module[target=L__self___net_0](args = (%l_inputs_,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,482] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_1 : [num_users=1] = call_module[target=L__self___net_1](args = (%l__self___net_0,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,482] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     return l__self___net_1
[rank0]:[2023-11-09 12:39:12,485] [7/0] torch._dynamo.backends.distributed: [DEBUG] run_node call_module, submod_1 got args tuple(T[torch.Size([20, 1000])])
[rank0]:[2023-11-09 12:39:12,488] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] 
[rank0]:[2023-11-09 12:39:12,488] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---submod_1 graph---
[rank0]:[2023-11-09 12:39:12,488] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] graph():
[rank0]:[2023-11-09 12:39:12,488] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_1 : [num_users=1] = placeholder[target=l__self___net_1]
[rank0]:[2023-11-09 12:39:12,488] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_2 : [num_users=1] = call_module[target=L__self___net_2](args = (%l__self___net_1,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,488] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_3 : [num_users=1] = call_module[target=L__self___net_3](args = (%l__self___net_2,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,488] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_4 : [num_users=1] = call_module[target=L__self___net_4](args = (%l__self___net_3,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,488] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_5 : [num_users=1] = call_module[target=L__self___net_5](args = (%l__self___net_4,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,488] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     return l__self___net_5
[rank0]:[2023-11-09 12:39:12,492] [7/0] torch._dynamo.backends.distributed: [DEBUG] run_node call_module, submod_2 got args tuple(T[torch.Size([20, 1000])])
[rank0]:[2023-11-09 12:39:12,494] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] 
[rank0]:[2023-11-09 12:39:12,494] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---submod_2 graph---
[rank0]:[2023-11-09 12:39:12,494] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] graph():
[rank0]:[2023-11-09 12:39:12,494] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_5 : [num_users=1] = placeholder[target=l__self___net_5]
[rank0]:[2023-11-09 12:39:12,494] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_6 : [num_users=1] = call_module[target=L__self___net_6](args = (%l__self___net_5,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,494] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_7 : [num_users=1] = call_module[target=L__self___net_7](args = (%l__self___net_6,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,494] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %_set_grad_enabled : [num_users=0] = call_function[target=torch._C._set_grad_enabled](args = (True,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,494] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     return l__self___net_7
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed: [DEBUG] run_node output, output got args tuple(tuple(T[torch.Size([20, 1000])]))
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] 
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---final graph---
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] graph():
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l_inputs_ : torch.Tensor [num_users=1] = placeholder[target=L_inputs_]
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %submod_0 : [num_users=1] = call_module[target=compiled_submod_0](args = (%l_inputs_,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %submod_1 : [num_users=1] = call_module[target=compiled_submod_1](args = (%submod_0,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %submod_2 : [num_users=1] = call_module[target=compiled_submod_2](args = (%submod_1,), kwargs = {})
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     return (submod_2,)
[rank0]:[2023-11-09 12:39:12,496] [7/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---------------
```


split modules fx graphs after:

```
[rank0]:[2023-11-09 12:34:31,396] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] 
[rank0]:[2023-11-09 12:34:31,396] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---submod_0 graph---
[rank0]:[2023-11-09 12:34:31,396] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] graph():
[rank0]:[2023-11-09 12:34:31,396] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l_inputs_ : torch.Tensor [num_users=1] = placeholder[target=l_inputs_]
[rank0]:[2023-11-09 12:34:31,396] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
[rank0]:[2023-11-09 12:34:31,396] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_0 : [num_users=1] = call_module[target=L__self___net_0](args = (%l_inputs_,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,396] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_1 : [num_users=1] = call_module[target=L__self___net_1](args = (%l__self___net_0,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,396] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,396] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     return l__self___net_1
[rank0]:[2023-11-09 12:34:31,401] [11/0] torch._dynamo.backends.distributed: [DEBUG] run_node call_module, submod_1 got args tuple(T[torch.Size([20, 1000])])
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] 
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---submod_1 graph---
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] graph():
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_1 : [num_users=1] = placeholder[target=l__self___net_1]
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_2 : [num_users=1] = call_module[target=L__self___net_2](args = (%l__self___net_1,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_3 : [num_users=1] = call_module[target=L__self___net_3](args = (%l__self___net_2,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_4 : [num_users=1] = call_module[target=L__self___net_4](args = (%l__self___net_3,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_5 : [num_users=1] = call_module[target=L__self___net_5](args = (%l__self___net_4,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,404] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     return l__self___net_5
[rank0]:[2023-11-09 12:34:31,413] [11/0] torch._dynamo.backends.distributed: [DEBUG] run_node call_module, submod_2 got args tuple(T[torch.Size([20, 1000])])
[rank0]:[2023-11-09 12:34:31,415] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] 
[rank0]:[2023-11-09 12:34:31,415] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---submod_2 graph---
[rank0]:[2023-11-09 12:34:31,415] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] graph():
[rank0]:[2023-11-09 12:34:31,415] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %_enter_autocast : [num_users=1] = call_function[target=torch.amp.autocast_mode._enter_autocast](args = (cuda, torch.float16, True, None), kwargs = {})
[rank0]:[2023-11-09 12:34:31,415] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_5 : [num_users=1] = placeholder[target=l__self___net_5]
[rank0]:[2023-11-09 12:34:31,415] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_6 : [num_users=1] = call_module[target=L__self___net_6](args = (%l__self___net_5,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,415] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___net_7 : [num_users=1] = call_module[target=L__self___net_7](args = (%l__self___net_6,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,415] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %_exit_autocast : [num_users=0] = call_function[target=torch.amp.autocast_mode._exit_autocast](args = (%_enter_autocast,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,415] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     return l__self___net_7
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed: [DEBUG] run_node output, output got args tuple(tuple(T[torch.Size([20, 1000])]))
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] 
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---final graph---
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] graph():
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l_inputs_ : torch.Tensor [num_users=1] = placeholder[target=L_inputs_]
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %submod_0 : [num_users=1] = call_module[target=compiled_submod_0](args = (%l_inputs_,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %submod_1 : [num_users=1] = call_module[target=compiled_submod_1](args = (%submod_0,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %submod_2 : [num_users=1] = call_module[target=compiled_submod_2](args = (%submod_1,), kwargs = {})
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     return (submod_2,)
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---------------
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] 
[rank0]:[2023-11-09 12:34:31,420] [11/0] torch._dynamo.output_graph: [INFO] Step 2: done compiler function compile_fn

jon-chuang commented:
Actually, I think we don't need this for set_grad? We only need it for autocast. Set grad mutates the global state, which is fine.

jon-chuang commented:
Replaced by: https://github.com/pytorch/pytorch/pull/113374
